core-api/src/main/java/org/silverpeas/core/calendar/ICalendarEventImportProcessor.java

```diff
@@ -0,0 +1,319 @@
+   * an importation of an iCalendar source content or a synchronization with an external calendar.
+   * The ony difference between an simple importation and a synchronization is that in the
+   * synchronization the calendar's synchronization date will be set for each event fetched from the
+   * iCalendar source; the behavior of the method doesn't change.
+   * </p>
+   * <p>
+   * The calendar must exist in Silverpeas otherwise an {@link IllegalArgumentException} is thrown.
+   * </p>
+   * @param calendar a calendar in Silverpeas.
+   * @param descriptor a descriptor of import with the iCalendar source from which the event will
+   * be parsed.
+   * @return the result of the import process.
+   * @throws ImportException exception thrown if the import fails.
+   */
+  protected ICalendarImportResult importInto(final Calendar calendar,
+      final ImportDescriptor descriptor) throws ImportException {
+    if (!calendar.isPersisted()) {
+      throw new IllegalArgumentException(
+          "The calendar " + calendar.getTitle() + " (id = " + calendar.getId() +
+              ") doesn't exist in Silverpeas");
+    }
+    final ICalendarImportResult importResult = new ICalendarImportResult();
+    iCalendarImporter.imports(descriptor, events -> Transaction.performInOne(() -> {
+      events.forEach(e -> {
+        CalendarEvent event = e.getLeft();
+        List<CalendarEventOccurrence> occurrences = e.getRight();
+        adjustSomeProperties(event);
+        EventOperationResult result = importEvent(calendar, event, occurrences);
+        result.created().ifPresent(ce -> importResult.incAdded());
+        result.updated().ifPresent(ue -> importResult.incUpdated());
+      });
+      return null;
+    }));
+    return importResult;
+  }
+
+  /**
+   * Imports into the specified calendar in Silverpeas the specified event with its modified
+   * occurrences.
+   * <p>
+   * If the event already exists in the calendar, then it is updated with the properties carried
+   * by the specified event. Otherwise this latter is planned into the calendar. All of the
+   * specified occurrences of the given event follows the same rule.
+   * </p>
+   * @param calendar the calendar into which the specified event and occurrences should be
+   * imported.
+   * @param event the event to plan or to update in the calendar.
+   * @param occurrences the occurrences of the events that are modified from the event or its
+   * recurrence rule and that have to be added into the calendar.
+   * @return the result of the importation. It has the event in the calendar if either the event
+   * was updated or its modified occurrences saved.
+   */
+  private EventOperationResult importEvent(final Calendar calendar, final CalendarEvent event,
+      final List<CalendarEventOccurrence> occurrences) {
+    OperationContext.fromCurrentRequester();
+    OperationContext.addStates(IMPORT);
+    try {
+      return Transaction.performInOne(() -> {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this lambda expression number of lines from 21 to at most 20. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS1188)